### PR TITLE
HIVE-24534 HIVE-24646: Strict checks between decimal/characters & bigint/doubles

### PR DESCRIPTION
**Pull request by zabetak**
### What changes were proposed in this pull request?
Throw an error when `hive.strict.checks.type.safety=true` and the query contains comparison between decimals and character or bigint and double types.

### Why are the changes needed?
To fail-fast and avoid unexpected query results. Examples in HIVE-24534.
Strict type checks between bigint and double types was implemented before but due to a bug it was not throwing an error but just a warning.

### Does this PR introduce _any_ user-facing change?
Queries relying on comparisons between decimal and character  or  bigint and double types will fail.

### How was this patch tested?
`mvn clean test -Dtest=TestNegativeCliDriver -Dqfile_regex="strict_type_decimal.*" -Dtest.output.overwrite`
`mvn test -Dtest=TestDecimalStringValidation`

### CODE REVIEW

**Comment by vineetgarg02** [comment]
LGTM (+1 pending tests)

**Comment by wangyum** [comment]
+1 for Throw an error when `hive.strict.checks.type.safety=true` and the query contains comparison between decimals and character types.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java [hunk 1]
```diff
@@ -801,6 +797,22 @@ private boolean unSafeCompareWithBigInt(TypeInfo otherTypeInfo, TypeInfo bigintC
...
     }
+    
+    private boolean isDecimalCharacterComparison(TypeInfo type1, TypeInfo type2) {
+      if(type1 instanceof PrimitiveTypeInfo && type2 instanceof PrimitiveTypeInfo) {
+        Set<Set<PrimitiveObjectInspector.PrimitiveCategory>> comparisons = new HashSet<>(3);
+        comparisons.add(EnumSet.of(
+            PrimitiveObjectInspector.PrimitiveCategory.DECIMAL, PrimitiveObjectInspector.PrimitiveCategory.CHAR));
+        comparisons.add(EnumSet.of(
+            PrimitiveObjectInspector.PrimitiveCategory.DECIMAL, PrimitiveObjectInspector.PrimitiveCategory.VARCHAR));
+        comparisons.add(EnumSet.of(
```
Comment (kasakrisz) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java:808]: These rules are seem to be constant. Can you please extract them?
```
private static final Set<Set<PrimitiveObjectInspector.PrimitiveCategory>> decimalCharacterComparisonRules =
  Collections.unmodifiableSet(new HashSet<>() {{add(...)...}})
```

**Review by kasakrisz — CHANGES_REQUESTED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java (hunk 1 above)
Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java]: Fixed in https://github.com/apache/hive/pull/1780/commits/ab7432b6f8af27755acc28b2034fd2d3870c0207. Overall there is more refactoring to be done but I want to keep it in a separate issue.

**Review by kasakrisz — APPROVED** [review]
+1 pending tests

**Comment by zabetak** [comment]
Close and reopen to re-trigger tests

**Comment by zabetak** [comment]
Updated PR to improve readability (https://github.com/apache/hive/pull/1780/commits/37e707705c6cc7c025bbe7fb1b67682062dc84c7) and at the same time address HIVE-24646.

**Comment by dengzhhu653** [comment]
+1

File: ql/src/test/results/clientpositive/llap/avrotblsjoin.q.out [hunk 2]
```diff
@@ -71,9 +71,9 @@ POSTHOOK: Input: _dummy_database@_dummy_table
 POSTHOOK: Output: default@table1_1
 POSTHOOK: Lineage: table1_1.col1 SCRIPT []
 POSTHOOK: Lineage: table1_1.col2 SCRIPT []
-WARNING: Comparing a bigint and a string may result in a loss of precision.
-WARNING: Comparing a bigint and a string may result in a loss of precision.
-WARNING: Comparing a bigint and a string may result in a loss of precision.
+WARNING: Comparing string and bigint may result in loss of information.
```
Comment (dengzhhu653) [inline ql/src/test/results/clientpositive/llap/avrotblsjoin.q.out:74]: Maybe `a loss of precision` is more user friendly compared to `information`.

File: ql/src/test/results/clientpositive/llap/partition_wise_fileformat2.q.out [hunk 3]
```diff
@@ -123,31 +123,6 @@ POSTHOOK: Input: default@partition_test_partitioned@dt=102
 		100
 		100
 		100
-238	val_238	102
```
Comment (dengzhhu653) [inline ql/src/test/results/clientpositive/llap/partition_wise_fileformat2.q.out:126]: the result of this qfile seems unrelated to the changes...

File: ql/src/test/results/clientpositive/llap/avrotblsjoin.q.out (hunk 2 above)
Comment (zabetak) [reply ql/src/test/results/clientpositive/llap/avrotblsjoin.q.out:74]: I changed the message `to loss of information` since precision may not make sense for certain types such as string. I don't have a strong preference though so I can restore the old message.

**Comment by zabetak** [comment]
Closing and reopen to retrigger tests.
